disaggregated_memory/memories/external_remote_memory_v2.py

```diff
@@ -0,0 +1,215 @@
+                    simulate the external memory. functional accesses will
+                    still be mirrored. By default, it is set to True.
+
+    * Notes *
+        To set a latency to access the remote memory for SST, the user has to
+        use the top-level runscript on SST-side to define the access latency
+        value. Noncoherent XBars are deprecated from this version of
+        ExternalRemoteMemory.
+    """
+
+    def __init__(
+        self,
+        size: "str" = None,
+        addr_range: AddrRange = None,
+        use_sst_sim: bool = True,
+    ):
+        """This class has to be initialized using either size or memory ranges.
+
+        Args:
+            size (str, optional): Size. Defaults to None.
+            addr_range (AddrRange, optional): Address Range. Defaults to None.
+            link_latency (Tick, optional): Additional latency. Defaults to None
+        """
+        super().__init__()
+
+        # TODO: marked for deletion
+        # We will create a non-coherent cross bar if the user wants to simulate
+        # latency for the remote memory links.
+        # self._xbar_required = False
+
+        # We setup the remote memory with size or address range. This allows us
+        # to quickly scale the setup with N nodes.
+        self._size = None
+
+        # We will either use size or addr range. This variable is used to keep
+        # a track of that.
+        self._set_using_addr_ranges = False
+
+        # The ExternalMemory is an AbstractMemory object that connects
+        # gem5 to SST as an external memory.
+        self.outgoing_request_bridge = ExternalMemory()
+
+        # Indicate whether the user is using SST or not.
+        self.outgoing_request_bridge.use_sst_sim = use_sst_sim
+
+        # TODO: The range and physical_address_ranges should have the same name
+        # to avoid confusion. The address map needs to be visible to the cores
+        # to use all types of CPUs including the O3 CPU.
+        self.outgoing_request_bridge.in_addr_map = True
+
+        # The user needs to provide either the size of the remote memory or the
+        # range of the remote memory.
+        if size is None and addr_range is None:
+            fatal("External memory needs to either have a size or a range!")
+        else:
+            if addr_range is not None:
+                self.outgoing_request_bridge.physical_address_ranges = [
+                    addr_range
+                ]
+                self._size = (
+                    self.outgoing_request_bridge.physical_address_ranges[
+                        0
+                    ].size()
+                )
+                self._set_using_addr_ranges = True
+            # The size will be setup in the board in case ranges are not given
+            # by the user.
+            else:
+                # There is no range information provided by the user. Depending
+                # upon the ISA, we have to fix the address.
```

Comment:
This is something that we need to discuss. Ranges should be set on the board imo. But we need a way to plug and play ExternalMemory with a range. This is because the remote memory may not be contiguous. So the question is: `Do the board or the memory need to know where the remote memory?`